### DISCUSSION TITLE
How to use an existing Logger?

### DISCUSSION THREAD

**Discussion by 0xTim** [awslabs/aws-sdk-swift#444] [Q&A] [answered]
All server-side Swift apps usually have an existing logger (e.g. in Vapor it's `app.logger`) that's already bootstrapped since it's done by the application on launch. How can we pass this to the SDK so it can use it to make logging work?

Bootstrapping another logger is also not possible since SwiftLog explicitly disallows this and will fatal error if you try

**Comment by kneekey23** [comment]
we have given you the ability to initialize the logger just once somewhere in your application. In order to initialize it you just need to call 
```
import ClientRuntime
SDKLoggingSystem.initialize(logLevel: .debug)
```

Alternatively, if you need finer grain control of instances of SwiftLog, you can call SDKLoggingSystem.add to control specific instances of the log handler. For example:

```
import ClientRuntime

SDKLoggingSystem.add(logHandlerFactory: S3ClientLogHandlerFactory(logLevel: .debug))
SDKLoggingSystem.add(logHandlerFactory: CRTClientEngineLogHandlerFactory(logLevel: .info))
SDKLoggingSystem.initialize()
```

Is this what you are asking?

**Comment by kneekey23** [comment, marked as answer]
ah I think what you are asking is if you have an existing logger and you do not want to use the SDK's logger. In that case, you need to conform your logger to the `LogAgent` protocol that exists in the `ClientRuntime` module.

```swift
public protocol LogAgent {
    /// name of the struct or class where the logger was instantiated from
    var name: String {get}
    
    /// Get or set the configured log level.
    var level: LogLevel {get set}
    
    /// This method is called when a `LogAgent` must emit a log message.
    ///
    /// - parameters:
    ///     - level: The `LogLevel` the message was logged at.
    ///     - message: The message to log.
    ///     - metadata: The metadata associated to this log message as a dictionary
    ///     - source: The source where the log message originated, for example the logging module.
    ///     - file: The file the log message was emitted from.
    ///     - function: The function the log line was emitted from.
    ///     - line: The line the log message was emitted from.
    func log(level: LogLevel,
             message: String,
             metadata: [String: String]?,
             source: String,
             file: String,
             function: String,
             line: UInt)
}
```

Then once you have done that you can pass it to an SDK client via the following code:
```swift
struct MyRuntimeConfiguration: SDKRuntimeConfiguration {
    var logger: LogAgent
    
    var retryer: SDKRetryer
    
    init(logger: LogAgent) throws {
        self.logger = logger //pass in your logger or instantiate it here
        self.retryer = try SDKRetryer()
    }
}

    let myRuntimeConfig = MyRuntimeConfiguration(logger: SwiftLogger(label: "SomeLogger")) //pass in your instance of the logger you already have
    let config = try S3Client.S3ClientConfiguration(region: "us-east-1", runtimeConfig: myRuntimeConfig)
    let s3Client = S3Client(config: config)
```

**Reply by 0xTim** [reply]
Yeah that's what I meant. Ok cool, I'll give that a go!

It would be great to either provide an option to pass in a `SwiftLog` instance directly to the `SDKRuntimeConfiguration` or add a public extension to make any `SwiftLog` logger automatically conform to `LogAgent`. This will make integrating the SDK with the rest of the server-side Swift ecosystem much easier

**Reply by kneekey23** [reply]
We can do both! Can you possibly create an issue for me describing both of these asks?

**Reply by 0xTim** [reply]
Course!
